src/librustc_codegen_llvm/builder.rs

```diff
@@ -1225,7 +1316,116 @@ impl Builder<'a, 'll, 'tcx> {
+                        llfn,
+                        args.as_ptr() as *const &llvm::Value,
+                        args.len() as c_uint,
+                        Some(&*(bundle.raw)), noname()
+                    )
+                }
+            }
+            None => unsafe {
+                llvm::LLVMRustBuildCall(
+                    self.llbuilder,
+                    llfn,
+                    args.as_ptr() as *const &llvm::Value,
+                    args.len() as c_uint,
+                    None, noname()
+                )
+            }
+        }
+    }
+
+    fn call_memcpy(
+        &mut self,
+        dst: &'ll Value,
+        src: &'ll Value,
+        n_bytes: &'ll Value,
+        align: Align,
+        flags: MemFlags,
+    ) {
+        if flags.contains(MemFlags::NONTEMPORAL) {
+            // HACK(nox): This is inefficient but there is no nontemporal memcpy.
+            let val = &self.load(src, align);
+            let ptr = &self.pointercast(dst, &self.cx().type_ptr_to(&self.cx().val_ty(val)));
+            &self.store_with_flags(val, ptr, align, flags);
+            return;
+        }
+        let cx = &self.cx();
+        let ptr_width = &self.cx().sess().target.target.target_pointer_width;
+        let key = format!("llvm.memcpy.p0i8.p0i8.i{}", ptr_width);
+        let memcpy = cx.get_intrinsic(&key);
+        let src_ptr = &self.pointercast(src, cx.type_i8p());
+        let dst_ptr = &self.pointercast(dst, cx.type_i8p());
+        let size = &self.intcast(n_bytes, cx.type_isize(), false);
+        let align = cx.const_i32(align.abi() as i32);
+        let volatile = cx.const_bool(flags.contains(MemFlags::VOLATILE));
+        &self.call(memcpy, &[dst_ptr, src_ptr, size, align, volatile], None);
+    }
+
+    fn call_memset(
+        &mut self,
+        ptr: &'ll Value,
+        fill_byte: &'ll Value,
+        size: &'ll Value,
+        align: &'ll Value,
+        volatile: bool,
+    ) -> &'ll Value {
+        let ptr_width = &self.cx().sess().target.target.target_pointer_width;
+        let intrinsic_key = format!("llvm.memset.p0i8.i{}", ptr_width);
+        let llintrinsicfn = &self.cx().get_intrinsic(&intrinsic_key);
+        let volatile = &self.cx().const_bool(volatile);
+        &self.call(llintrinsicfn, &[ptr, fill_byte, size, align, volatile], None)
+    }
+
+    fn zext(&mut self, val: &'ll Value, dest_ty: &'ll Type) -> &'ll Value {
+        self.count_insn("zext");
+        unsafe {
+            llvm::LLVMBuildZExt(self.llbuilder, val, dest_ty, noname())
+        }
+    }
+
+    fn struct_gep(&mut self, ptr: &'ll Value, idx: u64) -> &'ll Value {
+        self.count_insn("structgep");
+        assert_eq!(idx as c_uint as u64, idx);
+        unsafe {
+            llvm::LLVMBuildStructGEP(self.llbuilder, ptr, idx as c_uint, noname())
+        }
+    }
+
+    fn cx(&self) -> &'a CodegenCx<'ll, 'tcx, &'ll Value> {
+        &self.cx
+    }
+
+    fn delete_basic_block(&mut self, bb: &'ll BasicBlock) {
```

Comment:
This is **unsafe**! The trait method should be marked as such. Maybe we can find a way to avoid doing it at all.